File: cmd/myshell/cmds.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func handleExit(args []string) error {
	var (
		exitCode int
		err      error
	)

	if len(args) == 1 {
		exitCode, err = strconv.Atoi(args[0])
		if err != nil {
			return err
		}
	}

	os.Exit(exitCode)
	return nil
}

func handleEcho(args []string) error {
	if len(args) == 0 {
		fmt.Fprintln(os.Stdout)
		return nil
	}

	for i := 0; i < len(args)-1; i++ {
		fmt.Fprintf(os.Stdout, "%s ", args[i])
	}
	fmt.Fprintln(os.Stdout, args[len(args)-1])
	return nil
}

func locateCmd(cmd string) (string, bool) {
	path := os.Getenv("PATH")
	dirs := strings.Split(path, ":")
	for _, dir := range dirs {
		entries, err := os.ReadDir(dir)
		if err != nil {
			continue
		}

		for _, e := range entries {
			if e.IsDir() {
				continue
			}

			parts := strings.Split(e.Name(), ".")
			name := parts[0]
			if cmd == name {
				return fmt.Sprintf("%s/%s", dir, name), true
			}
		}
	}
	return "", false
}

func handleType(args []string) error {
	if len(args) != 1 {
		return nil
	}

	cmd := args[0]
	if _, ok := Cmds[cmd]; ok {
		fmt.Fprintf(os.Stderr, "%s is a shell builtin\n", cmd)
		return nil
	}

	if path, ok := locateCmd(cmd); ok {
		fmt.Fprintf(os.Stdout, "%s is %s\n", cmd, path)
		return nil
	}

	fmt.Fprintf(os.Stderr, "%s: not found\n", cmd)
	return nil
}

func handlePwd(args []string) error {
	dir, err := os.Getwd()
	if err != nil {
		return err
	}

	fmt.Fprintln(os.Stdout, dir)
	return nil
}

func handleCd(args []string) error {
	if len(args) == 0 {
		return nil
	}
	dir := args[0]
	if dir == "~" {
		dir = os.Getenv("HOME")
	}
	err := os.Chdir(dir)
	if err != nil {
		fmt.Fprintf(os.Stderr, "cd: %s: No such file or directory\n", dir)
	}
	return nil
}

File: src/cmds.rs
use std::env;
use std::io::{self, Write};
use std::process::{self, Command};

use crate::consts::CMDS;

//defining the type for command functions
pub type CmdFn = fn(Vec<String>) -> io::Result<()>;

pub fn handle_exit(args: Vec<String>) -> io::Result<()> {
    let exit_code = if args.is_empty() {
        0
    } else {
        args[0].parse().unwrap_or(0)
    };

    process::exit(exit_code);
}

pub fn handle_echo(args: Vec<String>) -> io::Result<()> {
    if args.is_empty() {
        println!();
        return Ok(());
    }

    let output = args.join(" ");
    println!("{}", output);
    Ok(())
}

pub fn locate_cmd(cmd: &str) -> Option<String> {
    if let Some(path) = env::var_os("PATH") {
        for dir in env::split_paths(&path) {
            let full_path = dir.join(cmd);
            if full_path.is_file() {
                return full_path.to_str().map(String::from);
            }
        }
    }
    None
}

pub fn handle_type(args: Vec<String>) -> io::Result<()> {
    if args.is_empty() {
        return Ok(());
    }

    let cmd = &args[0];
    //lock mutex before accessing it
    if let Ok(commands) = CMDS.lock() {
        if commands.contains_key(cmd) {
            eprintln!("{} is a shell builtin", cmd);
            return Ok(());
        }
    }

    if let Some(path) = locate_cmd(cmd) {
        println!("{} is {}", cmd, path);
        return Ok(());
    }

    eprintln!("{}: not found", cmd);
    Ok(())
}

pub fn handle_pwd(_args: Vec<String>) -> io::Result<()> {
    let current_dir = env::current_dir()?;
    println!("{}", current_dir.display());
    Ok(())
}

pub fn handle_cd(args: Vec<String>) -> io::Result<()> {
    if args.is_empty() {
        return Ok(());
    }

    let mut dir = args[0].clone();
    if dir == "~" {
        dir = env::var("HOME").unwrap_or_else(|_| String::from("~"));
    }

    if let Err(_) = env::set_current_dir(&dir) {
        eprintln!("cd: {}: No such file or directory", dir);
    }
    Ok(())
}

pub fn get_command(cmd: &str) -> Option<CmdFn> {
    if let Ok(commands) = CMDS.lock() {
        commands.get(cmd).copied()
    } else {
        None
    }
}

pub fn execute_external_command(cmd: &str, args: Vec<String>) -> io::Result<()> {
    match Command::new(cmd).args(args).output() {
        Ok(output) => {
            io::stdout().write_all(&output.stdout)?;
            io::stderr().write_all(&output.stderr)?;
            Ok(())
        }
        Err(err) => {
            eprintln!("{}", err);
            Ok(())
        }
    }
}
